src/webutil/tableState.js

```diff
@@ -0,0 +1,185 @@
+// @flow
+
+import {type PluckFn} from "../util/sortBy";
+import sortBy from "../util/sortBy";
+import cloneDeep from "lodash.clonedeep";
+import {useMemo, useState} from "react";
+
+export const SortOrders = {
+  ASC: "asc",
+  DESC: "desc",
+};
+type SortOrderTypes = "asc" | "desc";
+
+/**
+  A state manager for a React data table that provides encapsuled hooks for Sort, Pagination, and Filtering.
+ */
+export type TableState<T> = {|
+  /**
+    A stateful representation of the current page data.
+   */
+  +currentPage: $ReadOnlyArray<T>,
+  /**
+    A stateful representation of the length of the base data (unfiltered, unpaginated).
+   */
+  +length: number,
+  /**
+    A stateful representation of the current page index.
+   */
+  +pageIndex: number,
+  /**
+    A stateful representation of the number of rows per page.
+   */
+  +rowsPerPage: number,
+  /**
+    A stateful representation of the current sort name (the last sort name passed to setSortFn).
+   */
+  +sortName: string,
+  /**
+    A stateful representation of the current order, represented by "asc" or "desc".
+   */
+  +sortOrder: SortOrderTypes,
+  /**
+    A function that changes the currentPage to represent the indicated index.
+   */
+  +setPageIndex: (number) => void,
+  /**
+    A function that sets the rows per page and updates the pageIndex to an equivalent place.
+   */
+  +setRowsPerPage: (number) => void,
+  /**
+    A function that adds a filter to the map of filters or overwrites an existing filter that matches the key.
+    Every filter stored will be applied together as an AND operation (results will only include elements
+    that pass all filters).
+    Sets pageIndex to 0.
+
+    string: A key that represents the filter in the map.
+    function?: A function that returns false if an element should be excluded from the result. Null to remove the filter at the key.
+   */
+  +createOrUpdateFilterFn: (string, null | ((T) => boolean)) => void,
+  /**
+    A function that sorts the data. Current pageIndex is maintained.
+
+    string: A key representing the sort field. It is is used to toggle sort order and increase efficiency.
+    function?: A function that returns the property of the element that should be used in the sort comparator.
+   */
+  +setSortFn: (string, PluckFn<T>) => void,
+  /**
+    A function that overwrites the base data and reapplies current sort and filter functions.
+    Sets pageIndex to 0.
+   */
+  +setData: ($ReadOnlyArray<T>) => void,
+|};
+
+/**
+  A builder function for initializing a new TableState for use in a ReactNode.
+  It encapsulates Sort, Pagination, and Filtering.
+ */
+export function useTableState<T>(defaults?: {
```

Comment:
Maybe a name like `defaultOptions` or `defaultState` would be more descriptive

Reply:
yeah... I'm thinking `initialOptions`